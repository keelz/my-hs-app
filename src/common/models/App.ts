
export interface IComponentProps {
  className?: string | string[];
}
