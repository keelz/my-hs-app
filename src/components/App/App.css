.App {
  display: flex;
  flex-direction: column;
  flex-grow: 1;
  min-height: 100vh;
  text-align: center;
}

.App-logo {
  animation: App-logo-spin infinite 20s linear;
  height: 5rem;
}

.App-header {
  height: 3rem;
  padding: 1.25rem;
}

.App-title {
  font-size: 1.5rem;
}

.App-intro {
  font-size: large;
}

@keyframes App-logo-spin {
  from { transform: rotate(0deg); }
  to { transform: rotate(360deg); }
}
