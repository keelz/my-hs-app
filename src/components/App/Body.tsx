import * as React from 'react';
import AppRouter from './Router';
import { composeClassname } from '../../common/utils';
import { IComponentProps } from '../../common/models/app.model';

interface AppBodyProps extends IComponentProps {}

const AppBody: React.SFC<AppBodyProps> = props =>
  <section className={composeClassname('App-body')(props.className)}>
    <AppRouter />
  </section>;

export default AppBody;
