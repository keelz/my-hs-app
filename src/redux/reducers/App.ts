import { IActionType } from '../Types';
import { Dispatch } from 'redux';

/** MODELS */
export type AppState = {
  loaded: boolean;
};

/** FACTORIES */
export const appStateFactory = () => ({
  loaded: false,
});

/** ACTIONS */
export const APP_DID_LOAD = 'APP_DID_LOAD';
export const APP_SET_LOADED = 'APP_SET_LOADED';

/** ACTION CREATORS */
export const appDidLoadAction = () =>
  (dispatch: Dispatch) =>
  dispatch({ type: APP_DID_LOAD });

export const appSetLoadedAction = (loaded: boolean) =>
  (dispatch: Dispatch) =>
  dispatch({
    type: APP_SET_LOADED,
    payload: { loaded },
  });

/** REDUCER */
export default (state: AppState = appStateFactory(), action: IActionType<any>): AppState => {
  const { type } = action;
  switch (type) {
    case APP_SET_LOADED: return {
      loaded: action.payload.loaded,
    };
    default: return state;
  }
};
