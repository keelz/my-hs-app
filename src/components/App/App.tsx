import * as React from 'react';
import logo from '../../common/assets/img/logo.svg';
import './App.css';

export const App: React.SFC = () =>
  <div className="App">
    <header className="App-header">
      <img src={logo} className="App-logo" alt="logo" />
      <h1 className="App-title">Welcome to React</h1>
    </header>
    <p className="App-intro">
      To get started, edit <code>src/App.tsx</code> and save to reload.
    </p>
  </div>;

export default App;
