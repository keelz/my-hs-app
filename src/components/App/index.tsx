import * as React from 'react';
import App from './App';
import LoadingAnimation from '../LoadingAnimation';

// redux props.
type IAppStateProps = {
  loaded: boolean;
  appDidLoadAction: () => any;
};

class Index extends React.Component<IAppStateProps> {
  componentDidMount() {
    this.props.appDidLoadAction();
  }

  render() {
    return (
      <div className="App container">
        { this.props.loaded
          ? <App />
          : <LoadingAnimation />
        }
      </div>
    );
  }
}

export default Index;
