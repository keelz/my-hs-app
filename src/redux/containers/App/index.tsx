import { connect } from 'react-redux';
import { RootState } from '../../Types';
import App from '../../../components/App';
import { appDidLoadAction } from '../../reducers/App';

const mapStateToProps = (state: RootState) => ({
  loaded: state.App.loaded,
});

const mapDispatchToProps = {
  appDidLoadAction,
};

const connectedApp = connect(mapStateToProps, mapDispatchToProps)(App);

export default connectedApp;
