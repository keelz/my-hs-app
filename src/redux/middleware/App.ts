import { MiddlewareAPI, Dispatch, Middleware } from 'redux';
import hsJsonApi from '../../common/services/hsJsonApi';
import HSJSON from '../../common/constants/hsJson';
import { IActionType, IRootState } from '../Types';
import { cardsSetCardsAction, CARDS_SET_FILTER, cardsDeleteFilterAction } from '../reducers/Cards';
import {
  appSetLoadedAction,
  APP_DID_LOAD,
} from '../reducers/App';

/**
 * Define a redux middleware function.
 * https://redux.js.org/advanced/middleware
 */
const middleware = (api: MiddlewareAPI, state: IRootState) =>
(next: Dispatch<IActionType<any>>) =>
async (action: IActionType<any>) => {
  // Switch action type value to handle side effects.
  // Each case statement MUST return.
  // The Switch statement MUST have a default block that:
  // 1. calls the next: Dispatch<ActionType<any>> argument.
  // 2. returns.
  switch (action.type) {
    case APP_DID_LOAD:

      // **OPTIONAL**
      // Call the next argument with the action argument.
      // Calling the next argument with the action argument ensures that the
      // action received continues to the reducer.
      next(action);

      // **OPTIONAL**
      // complete asynchronous tasks synchronously.
      const response = await hsJsonApi.read(HSJSON.PATH.READ.LATEST);

      // **OPTIONAL**
      // dispatch side effects.
      cardsSetCardsAction(response.data)(api.dispatch);
      appSetLoadedAction(true)(api.dispatch);

      // **REQUIRED**
      // return void. each case statement that interprets the action type
      // correctly MUST return.
      // received.
      return;
    case CARDS_SET_FILTER:
      // capture state scope within anonymous function.
      (() => {
        const state = api.getState();
        const currentFilters = state.Cards.filters;
        const newFilters = action.payload.filters;
        if (!!newFilters[HSJSON.RESPONSE_PARAMS.COST]
          && !!currentFilters[HSJSON.RESPONSE_PARAMS.COST]
          && newFilters[HSJSON.RESPONSE_PARAMS.COST]
          === currentFilters[HSJSON.RESPONSE_PARAMS.COST]) {
          cardsDeleteFilterAction(HSJSON.RESPONSE_PARAMS.COST)(api.dispatch);
          return;
        }
        return next(action);
      })();
      return;
      // const currentFilters = state.Cards.filters;
      // const newFilters = action.payload.filters;
      // console.log(currentFilters, newFilters);

    // **REQUIRED**
    // Default case block MUST call next argument with action argument and return.
    default: return next(action);
  }
};

export default middleware as Middleware;
