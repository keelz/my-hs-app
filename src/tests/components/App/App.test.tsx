import * as React from 'react';
import * as enzyme from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import Index from '../../../components/App';
import { App } from '../../../components/App/App';

const defaultState = {
  loaded: true,
  appDidLoadAction: jest.fn(),
};

describe('App', () => {
  describe('snapshots', () => {
    it('renders App component correctly', () => {
      const tree = enzyme.shallow(<App />);
      expect(shallowToJson(tree)).toMatchSnapshot();
    });

    it('renders index component correctly when loaded is true', () => {
      const tree = enzyme.shallow(<Index {...defaultState} />);
      expect(shallowToJson(tree)).toMatchSnapshot();
    });

    it('renders index component correctly when loaded is false', () => {
      const tree = enzyme.shallow(<Index {...{ ...defaultState, loaded: false }} />);
      expect(shallowToJson(tree)).toMatchSnapshot();
    });
  });
});
