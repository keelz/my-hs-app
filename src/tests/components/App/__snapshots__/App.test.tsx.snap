// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App snapshots renders App component correctly 1`] = `
<BrowserRouter>
  <Connect(AppHeader) />
  <AppBody />
  <AppFooter />
</BrowserRouter>
`;

exports[`App snapshots renders index component correctly when loaded is false 1`] = `
<div
  className="App container"
>
  <LoadingAnimation />
</div>
`;

exports[`App snapshots renders index component correctly when loaded is true 1`] = `
<div
  className="App container"
>
  <App />
</div>
`;
